inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/support/AnnotationQueueOverviewDataProvider.java

```diff
@@ -300,50 +311,99 @@ public long getInProgressAmountForDocument(
-        SourceDocument document = null;
-        Random r = new Random();
+        AnnotatorState annotatorState = aAnnotationPageBase.getModelObject();
+
+        //First check for a documents that is in Progress for the user, and return this
+        for (AnnotationDocument annotationDocument: allAnnotationDocuments) {
+            //NULL CHECK due to listAnnotableDocuments from webanno
+            if (annotationDocument != null && annotationDocument.getState().equals(IN_PROGRESS)
+                //check for NAME here required, as comparing the annotationDocument with the
+                //current document will result in false
+                //This check is simply needed, because the transition from INPROGRESS to FINISHED in
+                //the DB is too slow
+                && !annotationDocument.getDocument().getName().equals(aCurrentAnno.getName())) {
+                //Found a document in progress
+                for (SourceDocument doc: data) {
+                    if (doc.getName().equals(annotationDocument.getName())) {
+                        return doc;
+                    }
+                }
+            }
+        }
 
-        while (document == null)
-        {
-            int i = r.nextInt(data.size() - 1);
-            //If the random chosen document wont surpass the amount of default number combining
-            // "inProgress" for the document + "finished" amount for the document + 1
-            if ((getInProgressAmountForDocument(data.
-                get(i)) +
-                getFinishedAmountForDocument(data.
-                    get(i)))
-                + 1 <= defaultAnnotations)
-            {
-                //If that was not the case, assign this document
-                document = data.get(r.nextInt(data.size() - 1));
+        //Nothing in Progress, so now return a random document, or create a new annotation document
+
+        //Get a random document
+        List<SourceDocument> srcList = entityManager.
+            createQuery(
+                "FROM SourceDocument " +
+                "WHERE project = :project " +
+                "ORDER BY rand()", SourceDocument.class)
+            .setParameter("project", annotatorState.getProject())
+            .getResultList();
+        for (SourceDocument src: srcList) {
+            //Look if there is a corresponding Annotation document
+            //If not, create a new one
+            AnnotationDocument anno = entityManager.
+                createQuery(
+                    "FROM AnnotationDocument " +
+                    "WHERE name = :sourceName " +
+                    "AND user = :currentUser ", AnnotationDocument.class)
+                .setParameter("sourceName", src.getName())
+                .setParameter("currentUser", annotatorState.getUser().getUsername())
+                .getResultList().stream().findFirst().orElse(null);
+
+            if (anno == null) {
+                createAnnotationDocument(src, annotatorState.getUser());
```

Comment:
Please do *not* create an annotation document! We have discussed this.

Reply:
@reckart reply in UKP messenger